firewood/src/proof.rs

```diff
@@ -113,24 +115,33 @@ impl<N: AsRef<[u8]> + Send> Proof<N> {
         let mut cur_hash = root_hash;
         let proofs_map = &self.0;
 
-        loop {
+        let value_node = loop {
             let cur_proof = proofs_map
                 .get(&cur_hash)
                 .ok_or(ProofError::ProofNodeMissing)?;
+
             let node = NodeType::decode(cur_proof.as_ref())?;
             let (sub_proof, traversed_nibbles) = locate_subproof(key_nibbles, node)?;
             key_nibbles = traversed_nibbles;
 
             cur_hash = match sub_proof {
                 // Return when reaching the end of the key.
-                Some(p) if key_nibbles.size_hint().0 == 0 => return Ok(Some(p.encoded)),
+                Some(p) if key_nibbles.size_hint().0 == 0 => break p.encoded,
                 // The trie doesn't contain the key.
                 Some(SubProof {
                     hash: Some(hash), ..
                 }) => hash,
                 _ => return Ok(None),
             };
-        }
+        };
+
+        let value = match NodeType::decode(&value_node) {
+            Ok(NodeType::Branch(branch)) => branch.value().as_ref().map(|v| v.to_vec()),
+            Ok(NodeType::Leaf(leaf)) => leaf.data().to_vec().into(),
+            _ => return Ok(value_node.into()),
```

Comment:
Weird case with extension nodes here... This will go away when extension nodes go away